[pruning][core][feature] Unit tests for structured pruning.

jcaip opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89776
* #89777
* #88436

Summary:

This PR contains unit tests for structured pruning `prune`.

Given a prepared structured pruning model (the result of pruner.prepare()),
 `prune` will resize the supported modules (currently conv2d or linear) by removing the masked weights

Previously, this functionality existed as a separate object in PrunedFuser (Karen's intern project).

The implementation of `prune` is in the subsequent PR.

I've refactored these tests a bit to be a part of test_structured_sparsifier instead of a new file, and pulled out some shared models into `common_pruning`.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

jcaip committed:
[pruning][core][feature] Unit tests and utility functions for structured pruning.

Summary:

This PR contains unit tests and some utility functions for structured pruning `prune`.

Given a prepared structured pruning model (the result of pruner.prepare()),
 `prune` will resize the supported modules (currently conv2d or linear) by removing the masked weights

Previously, this functionality existed as a separate object in PrunedFuser (Karen's intern project).

The implementation of `prune` is in the subsequent PR.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip commented on test/ao/sparsity/test_structured_sparsifier_prune.py:
```diff
@@ -0,0 +1,997 @@
+        module.parametrizations[tensor_name][0].mask[prune] = False
+
+
+class TestBaseStructuredSparsifierPrune(TestCase):
```
"test" code is below, everything else above are model definitions for the tests.

jcaip committed:
Update on "[pruning][core][feature] Unit tests and utility functions for structured pruning."


Summary:

This PR contains unit tests and some utility functions for structured pruning `prune`.

Given a prepared structured pruning model (the result of pruner.prepare()),
 `prune` will resize the supported modules (currently conv2d or linear) by removing the masked weights

Previously, this functionality existed as a separate object in PrunedFuser (Karen's intern project).

The implementation of `prune` is in the subsequent PR.

## Old Flow
```
original = Model()
pruner = ImplementedPruner({"sparsity_level": 0.5})
pruner.prepare(original, sparse_config)
pruner.enable_mask_update = True
pruner.step()
pruned_fuser = PrunedFuser(original, pruner)
fused_model = pruned_fuser.fuse()
```

## New Flow
```
original = Model()
pruner = ImplementedPruner({"sparsity_level": 0.5})
pruner.prepare(original, sparse_config)
pruner.enable_mask_update = True
pruner.step()
converted_model = pruner.prune()
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented:
I see changes to `torch/ao/pruning/_experimental/pruner/match_utils.py` but I don't see logic from this new file used anywhere, is this intended?

jcaip committed:
Update on "[pruning][core][feature] Unit tests and utility functions for structured pruning."


Summary:

This PR contains unit tests and some utility functions for structured pruning `prune`.

Given a prepared structured pruning model (the result of pruner.prepare()),
 `prune` will resize the supported modules (currently conv2d or linear) by removing the masked weights

Previously, this functionality existed as a separate object in PrunedFuser (Karen's intern project).

The implementation of `prune` is in the subsequent PR.

## Old Flow
```
original = Model()
pruner = ImplementedPruner({"sparsity_level": 0.5})
pruner.prepare(original, sparse_config)
pruner.enable_mask_update = True
pruner.step()
pruned_fuser = PrunedFuser(original, pruner)
fused_model = pruned_fuser.fuse()
```

## New Flow
```
original = Model()
pruner = ImplementedPruner({"sparsity_level": 0.5})
pruner.prepare(original, sparse_config)
pruner.enable_mask_update = True
pruner.step()
converted_model = pruner.prune()
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[pruning][core][feature] Unit tests and utility functions for structured pruning."


Summary:

This PR contains unit tests and some utility functions for structured pruning `prune`.

Given a prepared structured pruning model (the result of pruner.prepare()),
 `prune` will resize the supported modules (currently conv2d or linear) by removing the masked weights

Previously, this functionality existed as a separate object in PrunedFuser (Karen's intern project).

The implementation of `prune` is in the subsequent PR.

## Old Flow
```
original = Model()
pruner = ImplementedPruner({"sparsity_level": 0.5})
pruner.prepare(original, sparse_config)
pruner.enable_mask_update = True
pruner.step()
pruned_fuser = PrunedFuser(original, pruner)
fused_model = pruned_fuser.fuse()
```

## New Flow
```
original = Model()
pruner = ImplementedPruner({"sparsity_level": 0.5})
pruner.prepare(original, sparse_config)
pruner.enable_mask_update = True
pruner.step()
converted_model = pruner.prune()
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[pruning][core][feature] Unit tests for structured pruning."


Summary:

This PR contains unit tests for structured pruning `prune`.

Given a prepared structured pruning model (the result of pruner.prepare()),
 `prune` will resize the supported modules (currently conv2d or linear) by removing the masked weights

Previously, this functionality existed as a separate object in PrunedFuser (Karen's intern project).

The implementation of `prune` is in the subsequent PR.

I've refactored these tests a bit to be a part of test_structured_sparsifier instead of a new file, and pulled out some shared models into `common_pruning`.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[pruning][core][feature] Unit tests for structured pruning."


Summary:

This PR contains unit tests for structured pruning `prune`.

Given a prepared structured pruning model (the result of pruner.prepare()),
 `prune` will resize the supported modules (currently conv2d or linear) by removing the masked weights

Previously, this functionality existed as a separate object in PrunedFuser (Karen's intern project).

The implementation of `prune` is in the subsequent PR.

I've refactored these tests a bit to be a part of test_structured_sparsifier instead of a new file, and pulled out some shared models into `common_pruning`.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Commit:
Update on "[pruning][core][feature] Unit tests for structured pruning."


Summary:

This PR contains unit tests for structured pruning `prune`.

Given a prepared structured pruning model (the result of pruner.prepare()),
 `prune` will resize the supported modules (currently conv2d or linear) by removing the masked weights

Previously, this functionality existed as a separate object in PrunedFuser (Karen's intern project).

The implementation of `prune` is in the subsequent PR.

I've refactored these tests a bit to be a part of test_structured_sparsifier instead of a new file, and pulled out some shared models into `common_pruning`.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip commented:
Closing this as test are now in #89777
